File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
     vector<vector<int>> ans;
    void subseq(int k, int i, int sm, int n, vector<int>& temp){
        if(k == 0){
            if(sm == n){
                ans.push_back(temp);
            }
            return ;
        }
        if(i > 9)
            return ;
        sm += i;
        temp.push_back(i);
        subseq(k-1, i+1, sm, n, temp);
        sm -= i;
        temp.pop_back();
        subseq(k, i+1, sm, n, temp);
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        ans.clear();
        vector<int> temp;
        subseq(k, 1, 0, n, temp);

        return ans;
    }
};

File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    public void solve(int k, int n, int i, int sm , List<Integer> temp , List<List<Integer>> ans){
        if(k == 0){
            if(n == sm){
                ans.add(new ArrayList<>(temp));
            }
            return ;
        }
        if(i > 9)
            return;
        sm += i;
        temp.add(i);
        solve(k-1, n, i+1, sm, temp, ans);
        sm -= i;
        temp.remove(temp.size()-1);
        solve(k, n, i+1, sm, temp, ans);
    }
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();

        solve(k, n, 1, 0, temp, ans);
        return ans;
    }
}